src/quarto/game/game.go

```diff
@@ -76,8 +95,72 @@ func GetRemainingPiecesListFromState(state State) []int {
 // GetAllPiecesList generate a list of all pieces
 func GetAllPiecesList(state State) []int {
 	var piecesList []int
-	for i := 0; i < GridSize*GridSize; i++ {
+	size := GetGridSize(state)
+	for i := 0; i < size*size; i++ {
 		piecesList = append(piecesList, i+1)
 	}
 	return piecesList
 }
+
+// IsValid return false if the state is not acceptable
+func IsValid(state State) bool {
+	size := GetGridSize(state)
+	var piecesList = GetAllPiecesList(state)
+	if (len(state.Grid) != size) {
```

Comment:
Please split more `IsValid` function into multiple more readable and smaller functions.
It is still pretty big and difficult to digest.

But it can be done in another PR if you wish.